### ISSUE TITLE
Proximity sensor issue

### ISSUE THREAD

**Issue by vampik** [iNPUTmice/Conversations#3797]
#### General information

* **Version:** 2.8.7
* **Device:** Samsung J1 2016
* **Android Version:**  Android 5.1.1 (stock)
* **Server name:** self hosted
* **Server software:** ejabberd 20.04
* **Conversations source:** PlayStore

#### Steps to reproduce

1. Make or receive an audio call
2. Place the phone close to the ear (or just close the proximity sensor)

#### Expected result

The screen should be turned off.

#### Actual result

The screen is blinking: continuously turning on and off, on and off, on and off.. 

This seems to be a problem specific to this particular combination of device and software.
On another devices Conversations works correctly (screen is turning off during the call when being close to the ear).
Another applications on this device (including the stock dialer) works correctly too.

#### Debug output

Please post the output of adb logcat. The log should begin with the start of Conversations and include all the
steps it takes to reproduce the problem.

````
D/conversations( 3790): marina@messag.i.ng: delivered message to JingleRtpConnection <propose xmlns="urn:xmpp:jingle-message:0" id="1O+kPblcAaHBs+DhWJY8Zw"><description xmlns="urn:xmpp:jingle:apps:rtp:1" media="audio"/></propose>
D/conversations( 3790): marina@messag.i.ng: received session proposal from vamp@vampik.ru/Conversations.gXa0 for [audio]
D/conversations( 3790): marina@messag.i.ng: transitioned into PROPOSED
D/conversations( 3790): ForegroundService: on
D/conversations( 3790): marina@messag.i.ng: received call from vamp@vampik.ru/Conversations.gXa0. start ringing
D/conversations( 3790): ConversationsOverviewFragment.onPause()
D/conversations( 3790): intent action was accept
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): marina@messag.i.ng: transitioned into PROCEED
D/conversations( 3790): onJingleRtpConnectionUpdate(ACCEPTING_CALL)
D/conversations( 3790): ForegroundService: on
D/conversations( 3790): checking microphone availability took 408ms
D/conversations( 3790): marina@messag.i.ng: delivered message to JingleRtpConnection <accept xmlns="urn:xmpp:jingle-message:0" id="1O+kPblcAaHBs+DhWJY8Zw"/>
D/conversations( 3790): marina@messag.i.ng: unable to transition to accept because already in state=PROCEED
D/conversations( 3790): marina@messag.i.ng: packet delivered to JingleRtpConnection
D/conversations( 3790): processing session-init with 1 contents
D/conversations( 3790): marina@messag.i.ng: transitioned into SESSION_INITIALIZED_PRE_APPROVED
D/conversations( 3790): onJingleRtpConnectionUpdate(CONNECTING)
D/conversations( 3790): ForegroundService: on
D/conversations( 3790): marina@messag.i.ng: automatically accepting session-initiate
D/conversations( 3790): marina@messag.i.ng: packet delivered to JingleRtpConnection
D/conversations( 3790): marina@messag.i.ng: put ICE candidate on backlog
D/conversations( 3790): marina@messag.i.ng: packet delivered to JingleRtpConnection
D/conversations( 3790): marina@messag.i.ng: put ICE candidate on backlog
D/conversations( 3790): marina@messag.i.ng: packet delivered to JingleRtpConnection
D/conversations( 3790): marina@messag.i.ng: put ICE candidate on backlog
D/conversations( 3790): marina@messag.i.ng: packet delivered to JingleRtpConnection
D/conversations( 3790): marina@messag.i.ng: put ICE candidate on backlog
D/conversations( 3790): marina@messag.i.ng: discovered ICE Server: [stun:[54.38.159.159]:3478?transport=udp] [:] [TLS_CERT_POLICY_INSECURE_NO_CHECK] [] [null] [null]
D/conversations( 3790): marina@messag.i.ng: discovered ICE Server: [turn:[54.38.159.159]:3478?transport=udp] [1592476085:f693a8278348d8ae:vc5ZJMhf6zIuMSPIxqdyUCClDO0=] [TLS_CERT_POLICY_INSECURE_NO_CHECK] [] [null] [null]
D/conversations( 3790): ctor
D/conversations( 3790): create@[name=main, id=1]
D/conversations( 3790): ctor
D/conversations( 3790): AppRTCProximitySensor@[name=main, id=1]
D/conversations( 3790): setUseHardwareAcousticEchoCanceler(true) model=SM-J120F
D/conversations( 3790): defaultAudioDevice: EARPIECE
D/conversations( 3790): Android SDK: 22, Release: 5.1.1, Brand: samsung, Device: j1xlte, Id: LMY47X, Hardware: universal3475, Manufacturer: samsung, Model: SM-J120F, Product: j1xltejt
D/conversations( 3790): eu.siacs.conversations.services.AppRTCAudioManager.start()
D/conversations( 3790): Audio focus request granted for VOICE_CALL streams
D/conversations( 3790): start
D/conversations( 3790): BluetoothAdapter: enabled=false, state=OFF, name=Samsung Galaxy J1 (2016), address=68:E7:C2:7A:52:4E
D/conversations( 3790): HEADSET profile state: DISCONNECTED
D/conversations( 3790): Bluetooth proxy for headset profile has started
D/conversations( 3790): start done: BT state=HEADSET_UNAVAILABLE
D/conversations( 3790): --- updateAudioDeviceState: wired headset=false, BT state=HEADSET_UNAVAILABLE
D/conversations( 3790): Device status: available=[], selected=NONE, user selected=NONE
D/conversations( 3790): setAudioDeviceInternal(device=EARPIECE)
D/conversations( 3790): New device status: available=[EARPIECE, SPEAKER_PHONE], selected=EARPIECE
D/conversations( 3790): onAudioDeviceChanged in activity: selected:EARPIECE, available:[EARPIECE, SPEAKER_PHONE]
D/conversations( 3790): --- updateAudioDeviceState done
D/conversations( 3790): AudioManager started
D/conversations( 3790): onAudioDeviceChanged in activity: selected:EARPIECE, available:[EARPIECE, SPEAKER_PHONE]
D/conversations( 3790): no remote video tracks found
D/conversations( 3790): marina@messag.i.ng: added ICE candidate from back log audio:0:candidate:175950370 1 udp 2122262783 2001:470:26:3b1:b145:ac6:be96:6177 39308 typ host generation 0 ufrag 3j9c::UNKNOWN
D/conversations( 3790): marina@messag.i.ng: added ICE candidate from back log audio:0:candidate:2290633489 1 udp 2122194687 192.168.13.189 44935 typ host generation 0 ufrag 3j9c::UNKNOWN
D/conversations( 3790): marina@messag.i.ng: added ICE candidate from back log audio:0:candidate:842163049 1 udp 1685987071 95.24.233.254 44935 typ srflx raddr 192.168.13.189 rport 44935 generation 0 ufrag 3j9c::UNKNOWN
D/conversations( 3790): marina@messag.i.ng: added ICE candidate from back log audio:0:candidate:774123710 1 udp 41819903 54.38.159.159 55978 typ relay raddr 95.24.233.254 rport 44935 generation 0 ufrag 3j9c::UNKNOWN
D/conversations( 3790): marina@messag.i.ng: transitioned into SESSION_ACCEPTED
D/conversations( 3790): onJingleRtpConnectionUpdate(CONNECTING)
D/conversations( 3790): ForegroundService: on
D/conversations( 3790): <iq type="set"><jingle action="session-accept" xmlns="urn:xmpp:jingle:1" sid="1O+kPblcAaHBs+DhWJY8Zw"><group xmlns="urn:xmpp:jingle:apps:grouping:0" semantics="BUNDLE"><content name="audio"/></group><content name="audio" xmlns="urn:xmpp:jingle:1" creator="initiator"><description xmlns="urn:xmpp:jingle:apps:rtp:1" media="audio"><payload-type channels="2" name="opus" clockrate="48000" id="111" xmlns="urn:xmpp:jingle:apps:rtp:1"><parameter name="minptime" xmlns="urn:xmpp:jingle:apps:rtp:1" value="10"/><parameter name="useinbandfec" xmlns="urn:xmpp:jingle:apps:rtp:1" value="1"/><rtcp-fb type="transport-cc" xmlns="urn:xmpp:jingle:apps:rtp:rtcp-fb:0"/></payload-type><payload-type name="ISAC" id="103" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="16000"/><payload-type name="G722" id="9" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="8000"/><payload-type name="ILBC" id="102" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="8000"/><payload-type name="PCMU" id="0" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="8000"/><payload-type name="PCMA" id="8" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="8000"/><payload-type name="CN" id="105" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="16000"/><payload-type name="CN" id="13" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="8000"/><payload-type name="telephone-event" id="110" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="48000"/><payload-type name="telephone-event" id="113" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="16000"/><payload-type name="telephone-event" id="126" xmlns="urn:xmpp:jingle:apps:rtp:1" clockrate="8000"/><rtp-hdrext uri="urn:ietf:params:rtp-hdrext:ssrc-audio-level" id="1" xmlns="urn:xmpp:jingle:apps:rtp:rtp-hdrext:0"/><rtp-hdrext uri="http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time" id="2" xmlns="urn:xmpp:jingle:apps:rtp:rtp-hdrext:0"/><rtp-hdrext uri="http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01" id="3" xmlns="urn:xmpp:jingle:apps:rtp:rtp-hdrext:0"/><source xmlns="urn:xmpp:jingle:apps:rtp:ssma:0" ssrc="3731876595"><parameter name="cname" value="bN7yNossO506lVXJ"/><parameter name="msid" value="my-media-stream my-audio-track"/><parameter name="mslabel" value="my-media-stream"/><parameter name="label" value="my-audio-track"/></source><rtcp-mux/></description><transport ufrag="FFQp" pwd="X/65zXjGClMkD72elZJy3B7a" xmlns="urn:xmpp:jingle:transports:ice-udp:1"><fingerprint hash="sha-256" xmlns="urn:xmpp:jingle:apps:dtls:0" setup="active">59:0A:54:12:F7:89:FE:E8:7A:FE:DA:C4:21:1A:BE:14:F2:C1:38:12:7E:F9:7E:18:96:87:9A:7E:67:73:A7:10</fingerprint></transport></content></jingle></iq>
D/conversations( 3790): marina@messag.i.ng: PeerConnectionState changed to CONNECTING
D/conversations( 3790): onJingleRtpConnectionUpdate(CONNECTING)
D/conversations( 3790): sending candidate: audio:0:candidate:1029321089 1 udp 2122262783 2001:470:26:3b1:b900:b8eb:4f54:76bc 58146 typ host generation 0 ufrag FFQp network-id 4 network-cost 10::UNKNOWN
D/conversations( 3790): sending candidate: audio:0:candidate:3724228918 1 udp 2122194687 192.168.13.116 48105 typ host generation 0 ufrag FFQp network-id 3 network-cost 10::UNKNOWN
D/conversations( 3790): remote candidate selected: :-1:candidate:175950370 1 udp 2122262783 2001:470:26:3b1:b145:ac6:be96:6177 39308 typ host generation 0 ufrag 3j9c::UNKNOWN
D/conversations( 3790): local candidate selected: :-1:candidate:1029321089 1 udp 2122262783 2001:470:26:3b1:b900:b8eb:4f54:76bc 58146 typ host generation 0 ufrag FFQp network-id 4 network-cost 10::WIFI
D/conversations( 3790): marina@messag.i.ng: PeerConnectionState changed to CONNECTED
D/conversations( 3790): eu.siacs.conversations.xmpp.jingle.ToneManager.transition(CONNECTED)
D/conversations( 3790): eu.siacs.conversations.xmpp.jingle.ToneManager: do not configure audio manager because RTC has control
D/conversations( 3790): onJingleRtpConnectionUpdate(CONNECTED)
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): marina@messag.i.ng: PeerConnectionState changed to CLOSED
D/conversations( 3790): eu.siacs.conversations.xmpp.jingle.ToneManager.transition(ENDING_CALL)
D/conversations( 3790): eu.siacs.conversations.xmpp.jingle.ToneManager: do not configure audio manager because RTC has control
D/conversations( 3790): onJingleRtpConnectionUpdate(ENDING_CALL)
D/conversations( 3790): marina@messag.i.ng: transitioned into TERMINATED_SUCCESS
D/conversations( 3790): onJingleRtpConnectionUpdate(ENDED)
D/conversations( 3790): ForegroundService: on
D/conversations( 3790): <iq type="set"><jingle action="session-terminate" xmlns="urn:xmpp:jingle:1" sid="1O+kPblcAaHBs+DhWJY8Zw"><reason><success/></reason></jingle></iq>
D/conversations( 3790): eu.siacs.conversations.services.AppRTCAudioManager.stop()
D/conversations( 3790): stop: BT state=HEADSET_UNAVAILABLE
D/conversations( 3790): stopScoAudio: BT state=HEADSET_UNAVAILABLE, SCO is on: false
D/conversations( 3790): cancelTimer
D/conversations( 3790): stop done: BT state=UNINITIALIZED
D/conversations( 3790): Abandoned audio focus for VOICE_CALL streams
D/conversations( 3790): stop@[name=main, id=1]
D/conversations( 3790): eu.siacs.conversations.xmpp.jingle.ToneManager: putting AudioManager back into normal mode
D/conversations( 3790): ConversationsOverviewFragment.onStart()
D/conversations( 3790): ConversationsOverviewFragment.onResume()
D/conversations( 3790): releasing proximity wake lock

````

**Comment by iNPUTmice** [comment]
Yes. I think my Nexus 4 had the same issue. I don’t think there is anything Conversations can do about broken hardware or a broken OS.

**Comment by vampik** [comment]
The stock dialer works correctly. Another applications, for example, Google Duo, works correctly too. So this can be fixed for sure. This issue appears only in Conversations.
I think this is the relevant log lines:
```
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
D/conversations( 3790): acquiring proximity wake lock
D/conversations( 3790): releasing proximity wake lock
D/conversations( 3790): app switched into background
D/conversations( 3790): app switched into foreground
```

**Comment by iNPUTmice** [comment]
> So this can be fixed for sure.

I’m looking forward to your pull request.

**Comment by vampik** [comment]
> I’m looking forward to your pull request.

Well, I was hoping that you can investigate the issue and propose a fix which I can test.

I'll try to investigate this in the spare time.

**Comment by iNPUTmice** [comment]
For the record: 'Proximity wake lock' means that the operating system gets put in charge for turning the screen on and off (depending on the proximity sensor). When we leave the activity (the call screen) we want to tell the operating system to stop doing that. That's what the log lines `acquiring proximity wake lock` and `releasing proximity wake lock` refer to ('operating system start doing that' and 'operating system stop doing that' respectively)
The issue arises due to a bug in older Samsung phones (not sure up to which version but my Android 9 is fine) when Samsung stops (ends) the activity when turning off the screen. Samsung shouldn’t do that. That's wrong. From the perspective of the app there is no difference between Samsung turned of the screen and stopped the activity and the activity was legitimately stopped. However if the activity was legitimately stopped we do want to release the wake lock.
It's completely unclear to me how a work around for something like that would look like. Someone on stack overflow suggested to release the wakelock not on `onStop()` but on `onWindowsFocusChanged()` however that doesn’t work for rotations. because onWindowsFocusChanged() isn’t called for rotations. 
I’ve already wasted ~2 hours investigating this bug up to this point and I’m sick and tired wasting my life coming up with work around for Samsung bugs. Just stop buying garbage phones.

**Comment by vampik** [comment]
Sorry for the slight off-topic, but I'm trying to make my own build for testing on this Samsung phone and I don't know where to ask such questions.
1) What's the difference between +f**c**r and +f**s**r builds? I see the font included in +fcr builds, probably for emoji support, but I see no difference in UI. Both Samsung and Google keyboards offers their own emoji set and it doesn't depend on which build I install.
Furthermore I'm confused that the build from Google Play is called +p**c**r but doesn't include that font.
2) Am I right that I cannot make a working +pcr build myself? If I make a custom fixed build for Samsung it is useless for me without push notifications support.

**Comment by licaon-kter** [comment]
c means compat, includes the latest font, if s then used installed fonts, they might be the same, right on latest Android, but this works or 4.x and later.

**Comment by iNPUTmice** [comment]
Yes compat (c) vs system (s) is the font. However the play store variant downloads the font through Google Play which might fail.

It's probably best if you build fsd or fcd. You can’t have Google push without setting up a push server. But it is absolutely unnecessary. 3/4 of our users are on F-Droid and use Conversations without 'push' without any issue. Push probably doesn’t mean what you think it means.

Aside from that you basically just need that for testing. Once you have something that works and it has been merged upstream you can switch back to the Google build.

**Comment by vampik** [comment]
> Push probably doesn’t mean what you think it means.

I think it means battery saving, doesn't it?
That's the reason I've bought a SIP client with push support and the XMPP client with push support.
For example, a SIP client without push was hardly draining the battery, waking up the device every couple of minutes for communication with the server.

**Comment by licaon-kter** [comment]
There's no visible drain without push either. FYI

**Comment by vampik** [comment]
Well, the fix suggested on Stack Overflow works just fine for me. Screen rotation works fine too. If the device was rotated when the proximity sensor was closed, the screen rotates just a second after the proximity sensor was opene.

**Comment by vampik** [comment]
If you don't like this solution for some reason you may alter it to use onWindowFocusChanged() only on Samsung phones with OS <= 6 (my another Samsung with Android 7.1.1 works fine without fix).

**Comment by iNPUTmice** [comment]
Fixed in 203e248a14ed0d8351d59b027b133b5975db1725
